tests/components/mill/test_init.py

```diff
@@ -0,0 +1,126 @@
+"""Tests for Mill init."""
+
+from unittest.mock import patch
+
+import pytest
+
+from homeassistant.components import mill
+from homeassistant.exceptions import ConfigEntryNotReady
+
+from tests.common import MockConfigEntry, mock_coro
+
+
+async def test_setup_with_cloud_config(hass):
+    """Test setup of cloud config."""
+    mock_entry = MockConfigEntry(
+        domain=mill.DOMAIN,
+        data={
+            mill.CONF_USERNAME: "user",
+            mill.CONF_PASSWORD: "pswd",
+            mill.CONNECTION_TYPE: mill.CLOUD,
+        },
+    )
+    mock_entry.add_to_hass(hass)
+    with patch("mill.Mill.fetch_heater_and_sensor_data", return_value={}) as mock_fetch:
+        with patch("mill.Mill.connect", return_value=True) as mock_connect:
+            assert await mill.async_setup_entry(hass, mock_entry)
+    assert isinstance(
+        hass.data[mill.DOMAIN][mill.CLOUD]["user"].mill_data_connection, mill.Mill
+    )
+    assert len(mock_fetch.mock_calls) == 1
+    assert len(mock_connect.mock_calls) == 1
+
+
+async def test_setup_with_cloud_config_fails(hass):
+    """Test setup of cloud config."""
+    mock_entry = MockConfigEntry(
+        domain=mill.DOMAIN,
+        data={
+            mill.CONF_USERNAME: "user",
+            mill.CONF_PASSWORD: "pswd",
+            mill.CONNECTION_TYPE: mill.CLOUD,
+        },
+    )
+    mock_entry.add_to_hass(hass)
+    with patch("mill.Mill.connect", return_value=False):
+        with pytest.raises(ConfigEntryNotReady):
+            assert not await mill.async_setup_entry(hass, mock_entry)
+
+
+async def test_setup_with_old_cloud_config(hass):
+    """Test setup of old cloud config."""
+    mock_entry = MockConfigEntry(
+        domain=mill.DOMAIN,
+        data={
+            mill.CONF_USERNAME: "user",
+            mill.CONF_PASSWORD: "pswd",
+        },
+    )
+    mock_entry.add_to_hass(hass)
+    with patch("mill.Mill.fetch_heater_and_sensor_data", return_value={}):
+        with patch("mill.Mill.connect", return_value=True):
+            assert await mill.async_setup_entry(hass, mock_entry)
+
+    assert isinstance(
+        hass.data[mill.DOMAIN][mill.CLOUD]["user"].mill_data_connection, mill.Mill
```

Comment:
My impression from watching other PRs is that the best practice is not to interact with hass.data form the tests, but use some other details about the integration.  I think asserting the mock objects were invoked could be better aligned with that.

Reply:
(However, i think doing an assert on an empty hass data on unload is OK (e.g. like this test does
https://github.com/home-assistant/core/commit/458bc92124a349e0738036602c871a79505964de#diff-03b89207ac91dcf6ef3d8ff74ea8f8c5a9b18a9dbdc16e5fcbaf65aa789a3cceR11)